### PR TITLE
feat(common): Add support for DataType::Serial

### PR DESCRIPTION
**Pull request by shanicky**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

https://github.com/risingwavelabs/rfcs/pull/45

## Summary

This pull request introduces a new logical data type called `SERIAL` to the codebase. This new type is defined in  `proto/data.proto` files, and instantiated in `common/src/types/mod.rs`. Additionally, the code is updated to handle this new data type in various functions.

## Changes Made

- Added the `SERIAL` data type definition to `proto/data.proto`.
- Defined the `SERIAL` data type in `common/src/types/mod.rs`.
- Updated various functions to handle the `SERIAL` data type.
- Added support for the `SERIAL` data type in `common/src/hash/dispatcher.rs`.
- Mapped the `SERIAL` data type in `common/src/types/postgres_type.rs`.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
~- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/tests/sqlsmith/src/sql_gen/types.rs [hunk 1]
```diff
@@ -33,6 +33,7 @@ pub(super) fn data_type_to_ast_data_type(data_type: &DataType) -> AstDataType {
         DataType::Int16 => AstDataType::SmallInt,
         DataType::Int32 => AstDataType::Int,
         DataType::Int64 => AstDataType::BigInt,
+        DataType::Serial => AstDataType::Custom(vec!["SERIAL".into()].into()),
```
Comment (xiangjinwu) [inline src/tests/sqlsmith/src/sql_gen/types.rs:36]: This should be `unreachable`. `sqlsmith` should not generate a **user query** containing `SERIAL` data type word, which we do not have corresponding parsing logic in binder intentionally.

File: src/utils/pgwire/src/pg_extended.rs [hunk 2]
```diff
@@ -406,7 +406,7 @@ impl PreparedStatement {
                         .to_string(),
                     Format::Text => cstr_to_str(raw_param).unwrap().to_string(),
                 },
-                DataType::Int64 => {
+                DataType::Int64 | DataType::Serial => {
```
Comment (xiangjinwu) [inline src/utils/pgwire/src/pg_extended.rs:409]: Better use an error. We do not expect an input parameter of type `serial` as part of prepared statement.

File: src/utils/pgwire/src/pg_extended.rs [hunk 3]
```diff
@@ -542,6 +542,7 @@ impl PreparedStatement {
             match oid {
                 DataType::Boolean => params.push("false".to_string()),
                 DataType::Int64 => params.push("0::BIGINT".to_string()),
+                DataType::Serial => params.push("0::BIGINT".to_string()),
```
Comment (xiangjinwu) [inline src/utils/pgwire/src/pg_extended.rs:545]: same as above: drop support of `serial` in prepared statement

**Review by xiangjinwu — COMMENTED** [review]
lgtm. Left some suggestions to reject `serial` as input at certain places (except for array access `v1[idx]`).

BTW shall we customize the error message when user uses the `serial` word for a type?

Current:
```
dev=> create table t (v1 serial);
ERROR:  QueryError: Feature is not yet implemented: unsupported data type: serial
No tracking issue yet. Feel free to submit a feature request at https://github.com/risingwavelabs/risingwave/issues/new?labels=type%2Ffeature&template=feature_request.yml
```

Maybe we should give a more accurate message suggesting them not to use `serial`.

**Comment by kwannoel** [comment]
Wonder what is usage of `serial` for? Will you be adding some tests for its usage?

**Comment by kwannoel** [comment]
> lgtm. Left some suggestions to reject `serial` as input at certain places (except for array access `v1[idx]`).
> 
> BTW shall we customize the error message when user uses the `serial` word for a type?
> 
> Current:
> 
> ```
> dev=> create table t (v1 serial);
> ERROR:  QueryError: Feature is not yet implemented: unsupported data type: serial
> No tracking issue yet. Feel free to submit a feature request at https://github.com/risingwavelabs/risingwave/issues/new?labels=type%2Ffeature&template=feature_request.yml
> ```
> 
> Maybe we should give a more accurate message suggesting them not to use `serial`.

Screenshot 2023-03-14 at 3 51 48 PM

Seems like postgresql supports it, will we eventually support it too?

**Comment by shanicky** [comment]
> Wonder what is usage of `serial` for? Will you be adding some tests for its usage?

We currently use the serial type to implement the requirements in [rfc](https://github.com/risingwavelabs/rfcs/pull/45), the serial type is not visible to the user and is only used to implement row_id

File: src/frontend/src/binder/expr/mod.rs [hunk 4]
```diff
@@ -512,6 +512,13 @@ pub fn bind_data_type(data_type: &AstDataType) -> Result<DataType> {
                 "float8" => DataType::Float64,
                 "timestamptz" => DataType::Timestamptz,
                 "jsonb" => DataType::Jsonb,
+                "serial" => {
+                    return Err(ErrorCode::NotSupported(
+                        "Column type SERIAL is not supported".into(),
+                        "Please remove the SERIAL column".into(),
```
Comment (xiangjinwu) [inline src/frontend/src/binder/expr/mod.rs:518]: Maybe clearer message that user is not supposed to create a column of system type `serial`, or the `serial` type they see in table schema is not meant to be used as PostgreSQL `serial` (yet).

**Review by xiangjinwu — APPROVED** [review]
lgtm

**Pull request merged**